<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/Python-Notebook-Banners/Exercise.png"  style="display: block; margin-left: auto; margin-right: auto;";/>
</div>

# Exercise: Hypothesis testing


In this exercise, we will use hypothesis testing to make an inference about a population parameter based on sample data.

## Learning objectives

By the end of this train, you should be able to:
* Calculate a test statistic.
* Determine critical values based on a chosen significance level.
* Make a decision on whether to reject the null hypothesis.

## Exercises

Given a population where the average weight of a species of butterfly is believed to be 150 mg with a standard deviation of 20 mg, you collect a sample of 30 butterflies and find their average weight to be 145 mg. Use a 5% significance level to test the hypothesis that the population mean weight is 150 mg against the alternative hypothesis that it is not 150 mg (determining whether the observed average weight of the sample of butterflies significantly deviates from the presumed population average weight).

In the exercises below, we will develop different functions, each responsible for a specific part of the hypothesis-testing process.

### Import libraries and dataset

In [1]:
from scipy.stats import norm

### Exercise 1

State the null and alternative hypotheses.

In [ ]:
# State the null and alternative hypotheses
# H0: The mean weight of the population is 150 mg (μ = 150)
# H1: The mean weight of the population is not 150 mg (μ ≠ 150)

### Exercise 2

Write a function that calculates the test statistic (z-value) based on the sample mean, population mean, population standard deviation, and sample size.

The test statistic can be calculated using the formula for a one-sample z-test:

> z = (x̄ - μ) / (σ / √n)

where:

- `x̄` is the sample mean.
- `μ` is the population mean under the null hypothesis.
- `σ` is the population standard deviation.
- `n` is the sample size.

In [13]:
import numpy as np

def calculate_z_value(sample_mean, population_mean, population_std, sample_size):
    standard_error = population_std / np.sqrt(sample_size)
    z_score = (sample_mean - population_mean) / standard_error
    return z_score


# Given data
sample_mean = 155  # Sample mean weight in mg
population_mean = 150  # Population mean weight in mg
population_std = 10  # Population standard deviation in mg
sample_size = 30  # Sample size
z_score = calculate_z_value(sample_mean, population_mean, population_std, sample_size)
print(f"Calculated z-score: {z_score}")




Calculated z-score: 2.7386127875258306


### Exercise 3

Write a function that returns the lower and upper critical z-values for a two-sided test at a specific significance level.

In [14]:
# Function to return the lower and upper critical z-values for a two-tailed test at a given significance level
def critical_z_values(significance_level):
    lower_critical_z = norm.ppf(significance_level / 2)
    upper_critical_z = norm.ppf(1 - significance_level / 2)
    return lower_critical_z, upper_critical_z

lower_z, upper_z = critical_z_values(0.05)
print(f"Lower critical z-value: {lower_z}")
print(f"Upper critical z-value: {upper_z}")

Lower critical z-value: -1.9599639845400545
Upper critical z-value: 1.959963984540054


### Exercise 4

Write a function which makes a decision on whether to reject or not reject the null hypothesis based on the test statistic and critical values.

In [11]:
# Function to decide whether to reject the null hypothesis based on the calculated z-score and critical z-values
def decide_null_hypothesis(z_score, lower_critical_z, upper_critical_z):
    if z_score < lower_critical_z or z_score > upper_critical_z:
        return "Reject the null hypothesis"
    else:
        return "Fail to reject the null hypothesis"
    
decision = decide_null_hypothesis(z_score, lower_z, upper_z)
print(decision)

Reject the null hypothesis


### Exercise 5

Run the following code that puts together all the functions we have created in the previous exercises to perform the hypothesis test for the given butterfly population data.

> What does our decision mean in the context of the butterfly population?

In [15]:
# Given data
sample_mean = 145  # sample mean weight in mg
population_mean = 150  # hypothesised population mean weight in mg
population_std = 20  # population standard deviation in mg
n = 30  # sample size
alpha = 0.05  # significance level

# Calculate the z-value
z = calculate_z_value(sample_mean, population_mean, population_std, n)

# Get critical z-values
lower_z, upper_z = critical_z_values(alpha)

# Make a decision
decision = decide_null_hypothesis(z, lower_z, upper_z)

print(f"Test Statistic (z-value): {z}")
print(f"Critical z-value (lower): {lower_z}, Upper: {upper_z}")
print(f"Decision: {decision}")

Test Statistic (z-value): -1.3693063937629153
Critical z-value (lower): -1.9599639845400545, Upper: 1.959963984540054
Decision: Fail to reject the null hypothesis


## Solutions

### Exercise 1

- **Null hypothesis (H0):** The population mean weight of the species of butterfly is 150 mg, i.e. μ = 150 mg.

- **Alternative hypothesis (HA):** The population mean weight of the species of butterfly is not 150 mg, i.e. μ ≠ 150 mg.

### Exercise 2

In [ ]:
def calculate_z_value(sample_mean, population_mean, population_std, sample_size):
    
    # Calculate the z-value using the formula for a one-sample z-test
    z_value = (sample_mean - population_mean) / (population_std / (sample_size ** 0.5))
    
    return z_value

The `calculate_z_value` function takes four parameters: `sample_mean`, `population_mean`, `population_std`, and `sample_size`. It uses these to calculate the **z_value** as per the formula provided above.

The z-value, or z-score, is a test statistic that helps us evaluate the strength of the evidence against the null hypothesis. It determines how many standard deviations the sample mean is away from the population mean under the null hypothesis. 

### Exercise 3

In [ ]:
def critical_z_values(significance_level):
    
    # Calculate the lower critical z-value.
    lower_z = norm.ppf(significance_level / 2)
    
      # Calculate the upper critical z-value.
    upper_z = norm.ppf(1 - significance_level / 2)
    
    return lower_z, upper_z

The `critical_z_values` function takes the `significance_level` as a parameter which, using `norm.ppf` (percent point function), helps calculate:

- `lower_z`: This is the lower critical z-value that cuts off the lower tail of the normal distribution, leaving an area of significance_level / 2 to the left.
- `upper_z`: This is the upper critical z-value that cuts off the upper tail of the normal distribution, leaving an area of significance_level / 2 to the right.

These critical values are used to determine the rejection region for a two-sided hypothesis test. 

### Exercise 4

In [ ]:
def decide_null_hypothesis(z_value, lower_z, upper_z):
    if z_value < lower_z or z_value > upper_z:
        decision = "Reject null hypothesis"
    else:
        decision = "Do not reject null hypothesis"
    return decision

The `decide_null_hypothesis` function takes the parameters: `z_value`, `lower_z`, and `upper_z` calculated in the previous functions. 

We use them to make a decision regarding the null hypothesis in a two-sided hypothesis test:
- If the test statistic (z-value) is less than the lower critical value or greater than the upper critical value, the decision is to "Reject null hypothesis".
- Otherwise, if the z_value falls within the range defined by the lower and upper critical values, the decision is to "Do not reject null hypothesis".

### Exercise 5

In [ ]:
# Given data
sample_mean = 145  # sample mean weight in mg
population_mean = 150  # hypothesised population mean weight in mg
population_std = 20  # population standard deviation in mg
n = 30  # sample size
alpha = 0.05  # significance level

# Calculate the z-value
z = calculate_z_value(sample_mean, population_mean, population_std, n)

# Get critical z-values
lower_z, upper_z = critical_z_values(alpha)

# Make a decision
decision = decide_null_hypothesis(z, lower_z, upper_z)

print(f"Test Statistic (z-value): {z}")
print(f"Critical z-value (lower): {lower_z}, Upper: {upper_z}")
print(f"Decision: {decision}")

Since the **z-value** of `-1.37` is within the **critical values** of `-1.96` and `+1.96`, we **do not reject the null hypothesis**.

The decision to not reject the null hypothesis indicates that there isn't sufficient evidence at the 5% significance level to conclude that the population mean weight of the species of butterfly is different from 150 mg. This suggests that the observed sample mean weight of 145 mg can be attributed to sampling variability and does not provide strong evidence against the population mean being 150 mg, given the sample size and standard deviation specified.

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/refs/heads/master/ALX_banners/ALX_Navy.png"  style="width:140px";/>
</div>

In [16]:
import numpy as np

mu = 50          # population mean
sigma = 16       # population standard deviation
n = 64           # sample size

# Standard deviation of the sample mean (Standard Error)
std_error = sigma / np.sqrt(n)

print(f"Standard deviation of the sample mean: {std_error}")

Standard deviation of the sample mean: 2.0


In [17]:
import numpy as np
from scipy import stats

student_scores = np.array([55, 62, 67, 58, 63, 59, 61, 68, 60, 64])
mu = 60
x_bar = np.mean(student_scores)
s = np.std(student_scores, ddof=1)
n = len(student_scores)
alpha = 0.05

# T statistic
t_stat = (x_bar - mu) / (s / np.sqrt(n))

# One-tailed p-value depending on sign of t_stat
p_value = stats.t.cdf(t_stat, df=n-1) if t_stat < 0 else (1 - stats.t.cdf(t_stat, df=n-1))

# Two-tailed p-value
p_value *= 2

print("T statistic:", t_stat)
print("P-value:", p_value)

T statistic: 1.343501592923275
P-value: 0.21200020507950157


In [22]:
# Function to reject or fail to reject the null hypothesis
def decide_null_hypothesis(p_value, alpha):
    if p_value < alpha:
        return "Reject the null hypothesis"
    else:
        return "Fail to reject the null hypothesis"
    
decision = decide_null_hypothesis(p_value, alpha)

# Display mean score, significance level, and decision
print(f"Mean score: {x_bar}")
print(f"Significance level (alpha): {alpha}")
print(f"Decision: {decision}")


Mean score: 61.7
Significance level (alpha): 0.05
Decision: Fail to reject the null hypothesis


In [23]:
#Foodland shoppers have a mean grocery bill of R370 with a standard deviation of R100. 
# What is the probability that a sample of 40 Foodland shoppers will have a mean grocery bill of over R400?
mu = 370
sigma = 100 
n = 40
sample_mean = 400
standard_error = sigma / np.sqrt(n)
z_score = (sample_mean - mu) / standard_error
p_value = 1 - norm.cdf(z_score)
print(f"Z-score: {z_score}")
print(f"Probability: {p_value}")

Z-score: 1.8973665961010275
Probability: 0.028889785561798664


In [24]:
import math
from scipy.stats import poisson

lambda_val = 8
k = 6

# Using scipy
prob = poisson.pmf(k, lambda_val)
print(f"Probability: {prob:.4f}")

# Manual calculation
manual_prob = (math.exp(-lambda_val) * (lambda_val**k)) / math.factorial(k)
print(f"Manual Probability: {manual_prob:.4f}")

Probability: 0.1221
Manual Probability: 0.1221


In [25]:
# Given a population with a mean of 100 and a standard deviation of 15, you sample 30 items. 
# What is the probability that the sample mean will be less than 95? 
# Use Python's scipy.stats.norm module to calculate the probability.

population_mean = 100
population_std = 15
sample_size = 30
sample_mean = 95
standard_error = population_std / np.sqrt(sample_size)
z_score = (sample_mean - population_mean) / standard_error
probability = norm.cdf(z_score)
print(f"Z-score: {z_score}")
print(f"Probability: {probability}")


Z-score: -1.8257418583505538
Probability: 0.033944577430914495
